index.bs

```diff
@@ -2387,6 +2430,90 @@ Note: Invoking this method from a [=browsing context=] where the [=Web Authentic
 
 </div>
 
+### Deserialize Registration ceremony options - PublicKeyCredential's `getCreationOptionsFromJSON()` Method ### {#sctn-getCreationOptionsFromJSON}
+
+<div link-for-hint="WebAuthentication/getCreationOptionsFromJSON">
+
+[=[WRPS]=] use this method to convert JSON-compatible serialized representations of options for
+{{CredentialsContainer/create()|navigator.credentials.create()}} to {{PublicKeyCredentialCreationOptions}}.
+
+Upon invocation, the [=client=] must take the provided options and return an identically-structured object,
+making sure to decode to {{ArrayBuffer}} those {{DOMString}} properties that are encoded with [=base64url encoding=]
+but must be {{ArrayBuffer}} values according to the definition of {{PublicKeyCredentialCreationOptions}}.
+
+<xmp class="idl">
+    partial interface PublicKeyCredential {
+        static PublicKeyCredentialCreationOptions getCreationOptionsFromJSON(PublicKeyCredentialCreationOptionsJSON options);
+    };
+
+    dictionary PublicKeyCredentialCreationOptionsJSON {
+        required PublicKeyCredentialRpEntity                    rp;
+        required PublicKeyCredentialUserEntityJSON              user;
+        required Base64URLString                                challenge;
+        required sequence<PublicKeyCredentialParameters>        pubKeyCredParams;
+        unsigned long                                           timeout;
+        sequence<PublicKeyCredentialDescriptorJSON>             excludeCredentials = [];
+        AuthenticatorSelectionCriteria                          authenticatorSelection;
+        DOMString                                               attestation = "none";
+        AuthenticationExtensionsClientInputsJSON                extensions;
+    };
+
+    dictionary PublicKeyCredentialUserEntityJSON {
+        required Base64URLString        id;
+        required DOMString              name;
+        required DOMString              displayName;
+    };
+
+    dictionary PublicKeyCredentialDescriptorJSON {
+        required Base64URLString        id;
+        required DOMString              type;
+        sequence<DOMString>             transports;
+    };
+
+    dictionary AuthenticationExtensionsClientInputsJSON {
+        DOMString                                       appid;
+        DOMString                                       appidExclude;
+        boolean                                         uvm;
+        boolean                                         credProps;
+        AuthenticationExtensionsLargeBlobInputsJSON     largeBlob;
```

Comment:
A challenge here is related to the handling of extensions that may not be defined within WebAuthn itself. One of the characteristics of the current specification is that extensions might be defined outside of WebAuthn, as has already been done for the [credProtect extension in CTAP2](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension), and which is not recognised or present in this IDL, yet which an RP can include in credential creation options. How do you propose to handle that?

Reply:
I agree, this is especially tricky since the spec is written with the assumption that clients might just pass extension inputs through without any explicit extension processing. It seems to me that it's impossible for the `fromJSON` methods to tell which fields would need to be converted to `BufferSource`s without explicit support for the particular extension, so I'm not sure the `fromJSON` methods even _can_ be made compatible with the assumed pass-through mechanism.

Reply:
Maybe we should change the pass thru semantics of extensions since I don't believe there is any client that even honours that.

Reply:
Maybe, but I think it's been brought up before and met with objections. It's certainly far too big a decision to make here in this PR, we'd have to bring it up to the rest of the WG first, maybe with its own dedicated issue.

Reply:
Huh, I guess I never hit send on the reply I'd typed up here. In short I'm of the opinion that we only support extensions specifically defined in this spec - anything outside of the spec, like `credProtect`, can be "manually managed by the RP". 

At least that's one way I imagined we handle the existence of extensions defined outside this spec.

Reply:
I'm afraid that's not the case; [§9. WebAuthn Extensions](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#sctn-extensions) explicitly references an external registry of extensions:

>The IANA "WebAuthn Extension Identifiers" registry [[IANA-WebAuthn-Registries]](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#biblio-iana-webauthn-registries) established by [[RFC8809]](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#biblio-rfc8809) can be consulted for an up-to-date list of registered [WebAuthn Extensions](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#webauthn-extensions).

And an earlier paragraph states

>Clients wishing to support the widest possible range of extensions MAY choose to pass through any extensions that they do not recognize to authenticators, generating the [authenticator extension input](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#authenticator-extension-input) by simply encoding the [client extension input](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#client-extension-input) in CBOR. [...]

which is a normative declaration that this is a valid implementation choice supported by the spec. So I'd say we need to either be compatible with that provision or remove it, and removing it would require consensus from the rest of the working group.

Reply:
I've opened an issue to discuss dropping generic extension pass-through: #1730

But on second thought, a possible work-around to this issue is to specify a way to flag binary values. For example, MongoDB Extended JSON encodes binary values as `{ "$binary": "<base64>" }`. Or we could say that any value with a particular prefix in the property name be interpreted as binary - for example, `$BufferSource_base64url_challenge` would be base64url-decoded and renamed to `challenge` in the output.

It feels quite hacky and it's not well standardized, but it could enable this to be compatible with generic extension pass-through.

Reply:
This was resolved by defining `AuthenticationExtensionsClientInputsJSON` as an empty dictionary, and then leaving instructions for clients on how to handle any present values (this is for `.create()`, something similar is currently defined for `.get()` too):

> Upon invocation, the [client](http://localhost:8000/#client) MUST convert the [options](http://localhost:8000/#dom-publickeycredential-parsecreationoptionsfromjson-options-options) argument into a new, identically-structured [PublicKeyCredentialCreationOptions](http://localhost:8000/#dictdef-publickeycredentialcreationoptions) object, using [base64url encoding](http://localhost:8000/#base64url-encoding) to decode any [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString) attributes in [PublicKeyCredentialCreationOptionsJSON](http://localhost:8000/#dictdef-publickeycredentialcreationoptionsjson) that correspond to [buffer source type](https://webidl.spec.whatwg.org/#dfn-buffer-source-type) attributes in [PublicKeyCredentialCreationOptions](http://localhost:8000/#dictdef-publickeycredentialcreationoptions). This conversion MUST extend to any [client extension inputs](http://localhost:8000/#client-extension-input) processed by the [client](http://localhost:8000/#client).
> 
> [AuthenticationExtensionsClientInputsJSON](http://localhost:8000/#dictdef-authenticationextensionsclientinputsjson) MAY include extensions registered in the IANA "WebAuthn Extension Identifiers" registry [[IANA-WebAuthn-Registries]](http://localhost:8000/#biblio-iana-webauthn-registries) but not defined in [§ 9 WebAuthn Extensions](http://localhost:8000/#sctn-extensions).